### PR TITLE
fix: resolve latent bugs from the 2026-07-02 audit (F1–F11) + F0 test coverage

### PR DESCRIPTION
**Pull request by ericchansen**
## Summary

Resolves every confirmed finding from the 2026-07-02 latent-bug audit sweep. Each source fix ships with a regression test, and dead code is deleted or wired up rather than guarded (per the repo's alpha discipline). F0 was already fixed on `master` but had no test — this PR closes that gap so **all** audited bugs are covered.

## Fixes (F1–F11)

| # | Sev | Subsystem | Fix |
|---|-----|-----------|-----|
| F1 | P1 | `backends/mm/openmm.py` | Analytical-grad path dropped Urey–Bradley energy + gradients. `_create_diff_handle` now mirrors the harmonic-bond block for each UB 1-3 pair, restoring energy agreement and non-zero UB param grads. |
| F2 | P2 | `backends/mm/openmm.py` | Analytical-grad path omitted CMAP energy. Extracted a shared `_build_cmap_force` helper and added the CMAP force to the diff handle (energy-only; CMAP has no tunable params). |
| F3 | P1 | `io/mm3.py` | MM3 vdW section corrupted on every save. `_update_mm3_vdw_lines` now splices only the fixed-width numeric columns in place, leaving type/context/opt bytes untouched (byte-stable). |
| F4 | P2 | `io/mm3.py` | 4th–6th order torsions (`54` lines) mislabelled as periodicity 1/2/3. Now carry `ff_col` 4/5/6 with the correct alternating-sign phase (even n → 180°, odd n → 0°). |
| F5 | P2 | `optimizers/scipy_opt.py` | JaxLoss surrogate path froze `objective.n_eval`, so `n_evaluations` and the divergence guard saw stale values. Added surrogate telemetry feeding live call-count/score to the callback and report. |
| F6 | P2 | `optimizers/optax.py`, `jaxopt_opt.py` | Reported final scores in JaxLoss surrogate units, corrupting cross-stage comparisons. Both now re-evaluate the returned point with the true `ObjectiveFunction` before reporting (mirrors scipy). |
| F7 | P2 | `models/molecule.py`, `io/tinker.py` | Element resolution guessed from the atom-type label (`CO`→cobalt, `CA`→calcium). Now uses the authoritative element, normalised to canonical casing. |
| F8 | P3 | `models/hessian.py` | NumPy `invert_ts_curvature` didn't symmetrize before `eigh`, diverging from the JAX twin on asymmetric input. Now symmetrizes first. |
| F9 | P3 | `io/mm3.py` | Removed the dead, unguarded `float()` branch in the MM3 vdW loader. |
| F10 | P3 | `models/structure.py` | Removed the dead `atom.element is None` fallback in `format_coords` (the property never returns `None`). |
| F11 | P3 | `models/molecule.py` | `with_hessian` dropped cached `_torsions`/`_improper_torsions`; now deep-copied like `_bonds`/`_angles`. |

## Regression test coverage (all findings)

| Bug | Test |
|-----|------|
| F0 | `test/test_bench_composed.py` — AST guard asserting every `SystemData(...)` call in the script uses only valid dataclass fields |
| F1 | `test/test_urey_bradley.py::TestOpenMMUreyBradley` (energy agreement + FD) |
| F2 | `test/test_cmap.py::TestCmapOpenMM::test_energy_and_param_grad_includes_cmap` |
| F3 | `test/test_models.py::TestMm3VdwRoundTrip` |
| F4 | `test/test_models.py::TestMm3HigherOrderTorsion` |
| F5 | `test/test_jaxopt_optimizer.py::TestScipyJaxLossTelemetry` |
| F6 | `test/test_optax.py::TestOptaxFinalScoreUnits` + `test/test_jaxopt_optimizer.py::test_reported_scores_in_objective_units` |
| F7 | `test/test_models.py::TestStructureElementResolution` |
| F8 | `test/test_invert_ts_curvature_jax.py::test_parity_asymmetric_input` |
| F9 | `test/test_models.py::TestMm3VdwOptLineTolerated` |
| F10 | `test/test_models.py::TestStructureFormatCoords` |
| F11 | `test/test_models.py::TestMoleculeWithHessianCarriesTorsions` |

The F0 test was verified to **fail on the original buggy code** (`freq_ref=`) and **pass on the fix**, and it runs with no backend/GPU.

## Validation

- `ruff check q2mm/ test/ scripts/` — passed
- `ruff format --check q2mm test scripts examples` — 163 files clean
- Core suite (no backends): **754 passed, 262 skipped**
- Backend/JAX suites (OpenMM, optax, jaxopt, TS-curvature) — all passed locally
- Diff scanned for secrets/PII — clean

## Notes

- Two commits: the F1–F11 source fixes + tests, then the F0 test (its fix is already on `master`).
- `F0` fix landed earlier in `4129fe5`; this PR only adds its missing test.

Co-authored-by: Copilot App <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: q2mm/optimizers/scipy_opt.py [hunk 1]
```diff
@@ -685,16 +709,22 @@ def _make_callback(self, objective: ObjectiveFunction, initial_score: float) ->
         def callback(_xk: Any, *args: Any, **kwargs: Any) -> bool:
             """Log progress and trigger early stopping on sustained divergence."""
             nonlocal diverge_count
-            score = objective.history[-1] if objective.history else float("nan")
+            if telemetry is not None and telemetry.get("last_score") is not None:
+                score = telemetry["last_score"]
+                n = telemetry["n_eval"]
+                baseline = telemetry.get("initial") or initial_score
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/scipy_opt.py]: Fixed in b3797e9. Switched to an explicit `is not None` check so a legitimate surrogate baseline of 0.0 is honored instead of falling back to `initial_score` in ObjectiveFunction units.

File: q2mm/models/molecule.py [hunk 2]
```diff
@@ -427,15 +451,15 @@ def from_structure(
         coords = []
         for atom in structure.atoms:
             atom_label = atom.atom_type_name or atom.element or ""
-            symbols.append(_extract_element(atom_label))
+            symbols.append(_structure_atom_element(atom))
             atom_types.append(atom_label.strip() or _extract_element(atom_label))
             coords.append(atom.coords)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/models/molecule.py:473]: Fixed in b3797e9. Added a `_structure_atom_label` helper that reads `atom.element` through a try/except, so an atom with no derivable element yields an empty label instead of raising. One clarification: the existing `or` short-circuit already spared type-only atoms (they carry a type name, so `atom.element` was never reached) — the crash was only possible for atoms lacking both a type name and a valid atomic_num. Added a regression test covering that dummy-atom case.

File: q2mm/io/mm3.py [hunk 3]
```diff
@@ -156,6 +156,27 @@ def _parse_mm3_vdw_params(path: Path) -> list[VdwParam]:
...
+    """Overwrite one fixed-width numeric column, preserving all other bytes.
+
+    Returns *line* unchanged if the column falls outside the existing line
+    length (the field was not present in fixed position), so trailing MM3
+    columns — formal charge, context flag, opt descriptor — are never
+    disturbed.
+    """
+    if len(line) < start + width:
+        return line
+    return line[:start] + f"{value:{width}.4f}" + line[start + width :]
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/io/mm3.py]: Fixed in 851680a. `_splice_fixed` now formats the value first and, if it exceeds the column width, logs a warning and returns the line unchanged instead of shifting every trailing byte — so the byte-stability guarantee holds. Added a `TestSpliceFixed` regression covering the overflow case.
